message_ix/model/includes/period_parameter_assignment.gms

```diff
@@ -1,114 +1,122 @@
+if ( sum(year_all$( cat_year("firstmodelyear",year_all) ), 1 ),
+    first_period(year_all)$( cat_year("firstmodelyear",year_all) ) = yes ;
+else
+    first_period(year_all)$( ORD(year_all) eq 1 ) = yes ;
+) ;
+if ( sum(year_all$( cat_year("lastmodelyear",year_all) ), 1 ),
+    last_period(year_all)$( cat_year("lastmodelyear",year_all) ) = yes;
+else
+    last_period(year_all)$( ORD(year_all) = CARD(year_all) ) = yes ;
+) ;
+
+* dynamic sets for MACRO initialization and base periods
+macro_initial_period(year_all) = no ;
+macro_initial_period(year_all)$( cat_year("initializeyear_macro",year_all) ) = yes ;
+macro_base_period(year_all) = no ;
+macro_base_period(year_all)$( ORD(year_all) = sum(year_all2$( first_period(year_all2) ), ORD(year_all2) - 1 ) )  = yes ;
+
+* assign set of historical years, the model horizon and the MACRO hoizon (includes 'macro_base_period')
+historical(year_all)$( ORD(year_all) < sum(year_all2$cat_year("firstmodelyear",year_all2), ORD(year_all2) ) ) = yes ;
+model_horizon(year_all) = no ;
+model_horizon(year_all)$( ORD(year_all) >= sum(year_all2$first_period(year_all2), ORD(year_all2) )
+    AND ORD(year_all) <= sum(year_all2$last_period(year_all2), ORD(year_all2) ) ) = yes ;
+macro_horizon(year_all) = no ;
+macro_horizon(year_all)$macro_base_period(year_all) = yes;
+macro_horizon(year_all)$model_horizon(year_all) = yes;
+
+*----------------------------------------------------------------------------------------------------------------------*
+* assignment of (cumulative) discount factors over time                                                                *
+*----------------------------------------------------------------------------------------------------------------------*
+
+* simple method to compute discount factor (but this approach implicitly assumes a constant interest rate)
+*discountfactor(year_all) = POWER( 1 / ( 1+interestrate(year_all) ), sum(year_all2$( ORD(year_all2) < ORD(year_all) ),
+*    duration_period(year_all2) ) ) ;
+
+* compute per-year discount factor (using a recursive method) - set to 1 by default (interest rate = 0)
+df_year(year_all) = 1 ;
+
+* recursively compute the per-year discount factor
+loop(year_all$( ORD(year_all) > 1 ),
+    df_year(year_all) =
+        sum(year_all2$( seq_period(year_all2,year_all) ), df_year(year_all2)
+            * POWER( 1 / ( 1 + interestrate(year_all) ), duration_period(year_all) ) ) ;
+) ;
+
+* store the per-year discount factor for later use in the file 'MESSAGE/scaling_investment_costs.gms'
+*discountfactor('last_year') = sum(last_period, df_year(last_period) ) ;
+
+* multiply per-year discount factor by discounted period duration
+df_period(year_all) =
+    df_year(year_all) * (
+* multiply the per-year discount factor by the geometric series of over the duration of the period
+        ( ( POWER( 1 + interestrate(year_all) , duration_period(year_all) ) - 1 )
```

Comment:
In the current master branch the calculation is:

```
        ( ( 1 - POWER( 1 / ( 1 + interestrate(year_all) ), duration_period(year_all) ) )
            / ( 1 - 1 / ( 1 + interestrate(year_all) ) ) )$( interestrate(year_all) )
``` 

which seems substantively different than here

Reply:
*CRITICAL CHANGE*